ouroboros-network/src/Ouroboros/Network/Connections/Concurrent.hs

```diff
@@ -0,0 +1,257 @@
+        Nothing -> error "bug: entry vanished"
+        Just _  -> Nothing
+  in  state { connectionMap = Map.alter alteration identifier (connectionMap state) }
+
+-- | Generic concurrent connection manager in which at most one connection for
+-- a given identifier is allowed. New connections will re-use existing
+-- connections, and existing connections will be rejected if there is already
+-- one here.
+--
+-- For example, we could choose
+--   resource := Socket
+--   connectionId := exists sockType . (SockAddr sockType, SockAddr sockType)
+-- and this would guarantee at most one connection per socket address pair,
+-- when the sockType is IPv* and it's a TCP socket.
+-- However, for Unix domain sockets, this is no good, because unnamed sockets
+-- all have (==) SockAddr values. A richer connection identifier type would be
+-- needed, and any servers or clients calling into this concurrent connection
+-- manager with Unix domain sockets would be required to come up with unique
+-- identifiers for unnamed ones (probably by taking the file descriptor number).
+concurrent
+  :: forall connectionId resource request reject handle t .
+     ( Ord connectionId )
+  => (forall provenance .
+         Initiated provenance
+      -> connectionId
+      -> resource
+      -> request provenance
+      -> IO (Decision provenance reject handle)
+     )
+  -- ^ A callback to run for each connection. The `handle` gives an interface
+  -- to that connection, allowing for inspection and control etc. of whatever
+  -- it's doing.
+  -- If this throws an exception, it is re-thrown, so that calls to
+  -- `include` will get them.
+  -- Non-exceptional domain-specific reasons to reject a connection are given
+  -- in the rejection variant.
+  -> (Connections connectionId resource request (Reject reject) (Accept handle) IO -> IO t)
+  -> IO t
+concurrent withResource k = do
+  stateVar :: MVar (State connectionId handle) <- newMVar $ State
+    { connectionMap    = Map.empty
+    , connectionNumber = 0
+    }
+  let connections = Connections { include = includeOne stateVar }
+  k connections `finally` killThreads stateVar
+
+  where
+
+  -- Exception handling in here should ensure that if the handler is succesfully
+  -- created, then it ends up in the shared state.
+  --
+  -- FIXME this one definition is too complex. Try to factor it into simpler
+  -- pieces.
+  includeOne
+    :: MVar (State connectionId handle)
+    -> connectionId
+    -> Resource provenance IO resource
+    -> request provenance
+    -> IO (Types.Decision provenance (Reject reject) (Accept handle))
+  includeOne stateVar connId resource request = mask $ \restore -> modifyMVar stateVar $ \state ->
+    case Map.lookup connId (connectionMap state) of
+      Nothing   -> case resource of
+        Existing res closeResource -> do
+          -- User must ensure that the handler does not take very long to.
+          -- set up. It should just be creating shared state.
```

Comment:
Is the first sentence a stale comment?

Reply:
Yes indeed. It's relevant to the `New` case not the `Existing`. Will update.

Reply:
Oh actually on further consideration, it _is_ relevant, just confusing and not very helpful.